janus_collector/src/lib.rs

```diff
@@ -0,0 +1,1032 @@
+    pub fn with_collect_poll_backoff(mut self, backoff: ExponentialBackoff) -> CollectorParameters {
+        self.collect_poll_wait_parameters = backoff;
+        self
+    }
+
+    /// URL for collect requests.
+    fn collect_endpoint(&self) -> Result<Url, Error> {
+        Ok(self.leader_endpoint.join("collect")?)
+    }
+}
+
+pub fn default_http_client() -> Result<reqwest::Client, Error> {
+    Ok(reqwest::Client::builder()
+        .redirect(reqwest::redirect::Policy::none())
+        .user_agent(COLLECTOR_USER_AGENT)
+        .build()?)
+}
+
+/// Collector state related to a collect job that is in progress.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct CollectJob<P> {
+    /// The URL provided by the leader aggregator, where the collect response will be available
+    /// upon completion.
+    collect_job_url: Url,
+    /// The collect request's batch time interval.
+    batch_interval: Interval,
+    /// The aggregation parameter used in this collect request.
+    #[derivative(Debug = "ignore")]
+    aggregation_parameter: P,
+}
+
+impl<P> CollectJob<P> {
+    pub fn new(
+        collect_job_url: Url,
+        batch_interval: Interval,
+        aggregation_parameter: P,
+    ) -> CollectJob<P> {
+        CollectJob {
+            collect_job_url,
+            batch_interval,
+            aggregation_parameter,
+        }
+    }
+
+    pub fn collect_job_url(&self) -> &Url {
+        &self.collect_job_url
+    }
+
+    pub fn batch_interval(&self) -> Interval {
+        self.batch_interval
+    }
+
+    pub fn aggregation_parameter(&self) -> &P {
+        &self.aggregation_parameter
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+/// The result of a collect request poll operation. This will either provide the aggregate result
+/// or indicate that the collection is still being processed.
+pub enum PollResult<T> {
+    /// The aggregate result from a completed collect request.
+    AggregateResult(#[derivative(Debug = "ignore")] T),
+    /// The collect request is not yet ready. If present, the [`RetryAfter`] object is the time at
+    /// which the leader recommends retrying the request.
+    NextAttempt(Option<RetryAfter>),
+}
+
+/// A DAP collector.
+#[derive(Debug)]
+pub struct Collector<V: vdaf::Collector>
+where
+    for<'a> Vec<u8>: From<&'a <V as vdaf::Vdaf>::AggregateShare>,
+{
+    parameters: CollectorParameters,
+    vdaf_collector: V,
+    http_client: reqwest::Client,
+}
+
+impl<V: vdaf::Collector> Collector<V>
+where
+    for<'a> Vec<u8>: From<&'a V::AggregateShare>,
+{
+    /// Construct a new collector
+    pub fn new(
+        parameters: CollectorParameters,
+        vdaf_collector: V,
+        http_client: &reqwest::Client,
+    ) -> Collector<V> {
+        Collector {
+            parameters,
+            vdaf_collector,
+            http_client: http_client.clone(),
+        }
+    }
+
+    /// Send a collect request to the leader aggregator.
```

Comment:
Q. Right now, I think the expected usage of this type is:

1) Call `start_collection` to receive a `CollectJob`.
2) Either:
 a) Repeatedly call `poll_once` until the collect job is completed, or
 b) Call `poll_until_complete` to have the library handle retrying for you.

WDYT about a simpler API for collection: `collect`, which receives the same arguments as `start_collection`, and returns a future that resolves to the collection result (or an error). That is, it would effectively fuse the existing `start_collection` & `poll_until_complete` methods.

The upside is mostly in simplification -- the API would be simpler, and we'd no longer need to expose several of the helper types like `CollectJob` or `PollResult`.

I can think of a couple potential downsides, but IMO they don't break this idea. I wonder if I'm missing something else:
1) Users of this library can no longer pass in an arbitrary `CollectJob`. I'd argue that's OK: the only thing they could add to the collect job over what they're passing to `start_collection` is the collect job URI, and that is an "implementation deatil" of DAP from the perpsective of a collect library user.

2) Users of this library can no longer call `poll_once` to effectively interleave polling of multiple futures. I think that's OK: standard utilities like `try_join!` or `select!` exist to allow multiple futures to be driven concurrently.

(Putting things another way: what is the use case for `poll_once` being public? If there isn't one, I think simplifying things in this way makes sense.)

Reply:
(the current API setup also creates some tension in configuration: if the user only ever calls `poll_once`, the `collect_poll_wait_parameters` are effectively ignored/unused state)

Reply:
I think it might be worth having `start_collection` and `poll_once` exposed for the sake of users that want to persist the state of collect jobs that are in flight. (i.e. a long-lived service that needs to handle process restarts gracefully, or a distributed service that needs to pass that intermediate state between servers) I do like the idea of dropping `poll_until_complete` in favor of an end-to-end `collect`.

While Janus will hand out an already-created collect job URI if it receives a duplicate CollectReq, I don't think DAP requires that leaders do so. Thus I'd like to provide downstream users a way to resume polling an existing URI, rather than making them send a new CollectReq.

Reply:
Hmm -- I think users that wish to persist state could get away with persisting only the batch interval/aggregation parameter, and not the collect job URI. But the reason I think this is up to DAP spec interpretation:

> While Janus will hand out an already-created collect job URI if it receives a duplicate CollectReq, I don't think DAP requires that leaders do so. Thus I'd like to provide downstream users a way to resume polling an existing URI, rather than making them send a new CollectReq.

I think you are correct that Leaders are not required to return identical collect job URIs if they receive identical `CollectReq`s. However, all of the text around how aggregators validate repeated collections is phrased around the batch & aggregation parameter being collected, rather than the collect job URI. That is, if a collector sends two identical `CollectReq`s and receives two different collect job URIs, the correct interpretation of those URIs is as two "aliases" to the same collect job (i.e. batch & aggregation parameter), rather than as IDs to two separate collect jobs. IMO the latter interpretation does not fit with the DAP specification, as if these two `CollectReq`s are treated as separate collect jobs, I think we'd fall afoul of the following text:

> Check if the report has already been aggregated with this aggregation parameter. If this check fails, the input share MUST be marked as invalid with the error report-replayed.

... since treating the two `CollectReq`s as separate aggregations would effectively cause losing the first collect job to block reattempting collection. (and since the Collector updating its durable state is not transactional with the aggregators' state updates in handling `CollectReq`, treating two identical `CollectReq`s as resulting in separate aggregations would be another "data loss by design" scenario.)

All that said, the exact relation between "aggregations", "batches", "collect jobs", & "collect job URIs" might just be underspecified. Please correct me if your interpretation is different or I'm missing something.

Given the above, I don't see much value in exposing the collect job URI to the user -- a Collector that persists the state of collected data would do so indexed on `(batch_interval, aggregation_parameter)`. If the Collector crashed while waiting on a collection, it would figure out its state from what was last written to durable storage, then resend an identical `CollectReq` to resume collection (albeit with a potentially-different collect job URI). Indeed, even if the Collector wanted to store the collect job URI to avoid the need to send an identical `CollectReq`, if it crashed after receiving the URI but before storing it, it would need to send an identical `CollectReq` anyway. Since the gains to exposing the collect job URI are marginal, I suggest we simply don't expose it, and we remove `poll_once` in favor of `collect`.

(also, the details of the collect job URI might change eventually, e.g. with https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/278; not exposing the collect job URI to users of this library would likely let us do the necesssary updates without requiring a breaking change)